fx-src/FirefoxStorage.js

```diff
@@ -0,0 +1,227 @@
+
+  "updateData": `
+    UPDATE collection_data
+      SET record = :record
+        WHERE collection_name = :collection_name
+        AND record_id = :record_id`,
+
+  "deleteData": `
+    DELETE FROM collection_data
+      WHERE collection_name = :collection_name
+      AND record_id = :record_id`,
+
+  "saveLastModified": `
+    REPLACE INTO collection_metadata (collection_name, last_modified)
+      VALUES (:collection_name, :last_modified)`,
+
+  "getLastModified": `
+    SELECT last_modified
+      FROM collection_metadata
+        WHERE collection_name = :collection_name`,
+
+  "getRecord": `
+    SELECT record
+      FROM collection_data
+        WHERE collection_name = :collection_name
+        AND record_id = :record_id`,
+
+  "listRecords": `
+    SELECT record
+      FROM collection_data
+        WHERE collection_name = :collection_name`
+};
+
+const createStatements = ["createCollectionData",
+                          "createCollectionMetadata",
+                          "createCollectionDataRecordIdIndex"];
+
+export default class FirefoxAdapter extends BaseAdapter {
+  constructor(dbname) {
+    super();
+    this.dbname = dbname;
+
+    // make a promise to initialize
+    const self = this;
+    this.init = Task.spawn(function* () {
+      let connection;
+      try {
+        connection = yield self._doOpenConnection();
+        yield connection.executeTransaction(function* doSetup() {
+          let schema = yield (connection.getSchemaVersion());
+
+          if (schema == 0) {
+
+            for (let statementName of createStatements) {
+              yield (connection.execute(statements[statementName]));
+            }
+
+            yield (connection.setSchemaVersion(1));
+          } else if (schema != 1) {
+            throw new Error("Unknown database schema: " + schema);
+          }
+        });
+        self.initialized = true;
+      } finally {
+        if (connection) {
+          connection.close();
+        }
+      }
+    });
+  }
+
+  _doOpenConnection() {
+    const opts = { path: "kinto.sqlite", sharedMemoryCache: false }
+    return Sqlite.openConnection(opts);
+  }
+
+  _doTransaction(statement, params){
+    const self = this;
+    return Task.spawn(function* (){
+      // get a connection
+      let connection = yield self._doOpenConnection();
+
+      // ensure we're initialized
+      yield self.init;
+
+      //execute the transaction
+      try {
+        return yield connection.executeCached(statement, params);
+      }
+      finally {
+        // ensure the connection is closed
+        connection.close();
+      }
+    });
+  }
+
+  clear() {
+    const params = {collection_name: this.dbname};
+    return this._doTransaction(statements.clearData, params);
+  }
+
+  create(record) {
+    const params = {
+      collection_name: this.dbname,
+      record_id: record.id,
+      record: JSON.stringify(record)
+    };
+    return this._doTransaction(statements.createData, params).then(() => {
```

Comment:
Paging @n1k0 for that. I think the adapter API needs explicit `open` and `close` that are called at the start and end of the expected lifecycle of this object.

The indexeddb version of this code just opens once in the constructor and leaves the connection open until the browser kills it.